Feature: Implement VotesERC20LockableV1 Contract

adamgall opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/243

Fixes [ENG-812](https://linear.app/decent-labs/issue/ENG-812/implement-lockable-and-mintable-erc20-votes-token-voteserc20lockablev1)

**Summary:**

This PR introduces a new ERC20 token contract, `VotesERC20LockableV1`, which extends the functionality of `VotesERC20V1` to include lockable transfers and minting capabilities. It also includes the necessary interfaces and comprehensive tests for the new contract.

**Changes:**

*   **New Contract `VotesERC20LockableV1.sol`:**
    *   Added `contracts/deployables/erc20/VotesERC20LockableV1.sol`.
    *   This contract inherits from `VotesERC20V1` and implements `ILockableV1` and `IMintableV1`.
    *   It allows the owner to lock and unlock token transfers.
    *   Whitelisted addresses can transfer tokens even when the contract is locked.
    *   The owner can mint new tokens.
    *   Includes an `isTransferable` modifier to control token movements based on lock status and whitelist.
    *   Implements `ERC165` for interface detection and includes a `getVersion()` function.
*   **New Interface `ILockableV1.sol`:**
    *   Added `contracts/interfaces/decent/deployables/ILockableV1.sol`.
    *   Defines the interface for contracts that support locking of transfers and whitelisting addresses.
    *   Includes events `Locked` and `Whitelisted`, and errors `IsLocked` and `CannotSwitchLockState`.
*   **New Interface `IMintableV1.sol`:**
    *   Added `contracts/interfaces/decent/deployables/IMintableV1.sol`.
    *   Defines the interface for contracts that support minting new tokens.
*   **New Test Suite `VotesERC20LockableV1.test.ts`:**
    *   Added `test/deployables/erc20/VotesERC20LockableV1.test.ts`.
    *   Provides comprehensive test coverage for `VotesERC20LockableV1`, including:
        *   Initialization with locked and unlocked states.
        *   Locking and unlocking functionality by owner and non-owner.
        *   Whitelisting and removing addresses from the whitelist.
        *   Token transfers (`transfer` and `transferFrom`) under various lock and whitelist conditions.
        *   Token minting by the owner when locked and unlocked.
        *   Version checking.
        *   ERC165 interface support.
        *   UUPS upgradeability.

**Motivation:**

*   To provide a flexible ERC20 token that can have its transfers restricted by a central owner.
*   To allow specific addresses (e.g., exchanges, bridges) to bypass transfer restrictions via a whitelist.
*   To enable the owner to mint new tokens as needed.
*   To ensure the contract is robust, well-tested, and upgradeable.

dacebt reviewed (approved):
Nice looks like the tests cover all the scenarios I could think of. DAO will be owner 99%, this allows for the DAO to treat as normal, holders can not transfer unless given whitelisted permission to manage their own token otherwise has to remain locked until a vote to change it. 

I don't remember the conclusion of the discussion but is this to replace the standard ERC20VotesV1 or we offering both versions?

mudrila commented on test/deployables/erc20/VotesERC20LockableV1.test.ts:
wen `foundry` / `forge`

mudrila commented on test/deployables/erc20/VotesERC20LockableV1.test.ts:
```diff
@@ -0,0 +1,826 @@
+
+  describe('Initialization', () => {
+    describe('Locked Behavior', () => {
+      it('should be be locked on deployment', async () => {
```
Guess it might be confusing when test name are nearly equal but describing different behaviour without mentioning what changes in the parameters. WDYT about changing test case name to be more descriptive?

adamgall replied:
good call, updated

mudrila reviewed (approved):
Test cases are just state of the art!
Have one minor suggestion about description for 2 test cases, but absolutely cool w/ not changing them

adamgall commented:
> I don't remember the conclusion of the discussion but is this to replace the standard ERC20VotesV1 or we offering both versions?

In the UI right now we offer both versions

adamgall committed:
Feature: Implement VotesERC20LockableV1 Contract

**Summary:**

This PR introduces a new ERC20 token contract, `VotesERC20LockableV1`, which extends the functionality of `VotesERC20V1` to include lockable transfers and minting capabilities. It also includes the necessary interfaces and comprehensive tests for the new contract.

**Changes:**

*   **New Contract `VotesERC20LockableV1.sol`:**
    *   Added `contracts/deployables/erc20/VotesERC20LockableV1.sol`.
    *   This contract inherits from `VotesERC20V1` and implements `ILockableV1` and `IMintableV1`.
    *   It allows the owner to lock and unlock token transfers.
    *   Whitelisted addresses can transfer tokens even when the contract is locked.
    *   The owner can mint new tokens.
    *   Includes an `isTransferable` modifier to control token movements based on lock status and whitelist.
    *   Implements `ERC165` for interface detection and includes a `getVersion()` function.
*   **New Interface `ILockableV1.sol`:**
    *   Added `contracts/interfaces/decent/deployables/ILockableV1.sol`.
    *   Defines the interface for contracts that support locking of transfers and whitelisting addresses.
    *   Includes events `Locked` and `Whitelisted`, and errors `IsLocked` and `CannotSwitchLockState`.
*   **New Interface `IMintableV1.sol`:**
    *   Added `contracts/interfaces/decent/deployables/IMintableV1.sol`.
    *   Defines the interface for contracts that support minting new tokens.
*   **New Test Suite `VotesERC20LockableV1.test.ts`:**
    *   Added `test/deployables/erc20/VotesERC20LockableV1.test.ts`.
    *   Provides comprehensive test coverage for `VotesERC20LockableV1`, including:
        *   Initialization with locked and unlocked states.
        *   Locking and unlocking functionality by owner and non-owner.
        *   Whitelisting and removing addresses from the whitelist.
        *   Token transfers (`transfer` and `transferFrom`) under various lock and whitelist conditions.
        *   Token minting by the owner when locked and unlocked.
        *   Version checking.
        *   ERC165 interface support.
        *   UUPS upgradeability.

**Motivation:**

*   To provide a flexible ERC20 token that can have its transfers restricted by a central owner.
*   To allow specific addresses (e.g., exchanges, bridges) to bypass transfer restrictions via a whitelist.
*   To enable the owner to mint new tokens as needed.
*   To ensure the contract is robust, well-tested, and upgradeable.

adamgall committed:
Update test descriptions for VotesERC20LockableV1 to clarify locked and unlocked states

**Changes:**
- Modified test case descriptions in `VotesERC20LockableV1.test.ts` to specify conditions for locked and unlocked states during deployment.

**Motivation:**
- To improve clarity and understanding of test cases related to the initialization behavior of the `VotesERC20LockableV1` contract.
